processor/src/main/java/org/mapstruct/ap/internal/util/SpecificCompilerWorkarounds.java

```diff
@@ -115,4 +135,161 @@ public static TypeElement replaceTypeElementIfNecessary(Elements elementUtils, T
         }
         return element;
     }
+
+    public static TypeMirror asMemberOf(DeclaredType containing, Element element, Types typeUtils,
+                                        Elements elementUtils) {
+        try {
+            // quick and dirty, until implementation below is more sound
+            return typeUtils.asMemberOf( containing, element );
+        }
+        catch ( IllegalArgumentException e ) {
+            TypeMirror eclipseWorkaround =
+                EclipseEnvironment.asMemberOf( containing, element, typeUtils, elementUtils );
+
+            if ( null != eclipseWorkaround ) {
+                return eclipseWorkaround;
+            }
+
+            throw e;
+        }
+    }
+
+    private static final class EclipseEnvironment {
```

Comment:
Well, it is only loaded if an IllegalArgumentException is thrown. But nevertheless, I'll improve the error handling anyway, so that in any case we get to know the `element` and `containingElement` that the method was called with. Plus the stacktrace, of course.